### PR TITLE
[Cross Client Batching - 1b] | Wiring TimeLock Batch endpoints

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
Wiring the multi client endpoints through TimeLockAgent

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added integration sanity test

**Where should we start reviewing?**:
MultiClientConjureTimelockResource.java

**Priority (whenever / two weeks / yesterday)**:
Would be good to do before tomorrow EOD

### CODE REVIEW

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java [hunk 1]
```diff
@@ -484,6 +487,18 @@ public void stressTestForPaxosEndpoints() {
...
+    @Test
+    public void leaderRespondsToMultiClientRequests() {
+        MultiClientConjureTimelockService multiClientConjureTimelockService =
+                cluster.currentLeaderFor(client.namespace()).multiClientService();
+        Set<String> expectedNamespaces = ImmutableSet.of("alpha", "beta");
+        List<NamespacedLeaderTime> leaderTimes =
+                multiClientConjureTimelockService.leaderTimes(AuthHeader.valueOf("Bearer omitted"), expectedNamespaces);
+        Set<String> namespaces =
+                leaderTimes.stream().map(NamespacedLeaderTime::getNamespace).collect(Collectors.toSet());
+        assertThat(namespaces).hasSameElementsAs(expectedNamespaces);
```
Comment (Jolyon-S) [inline timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:499]: I don't know leader time (and/or the internals of Timelock) all that well - would leader times from different namespaces have different leadership IDs, or the same? If you expect one or the other, can you verify that in this test?

**Review by Jolyon-S — COMMENTED** [review]
Just one comment on the test. I haven't been keeping up with the CCB stuff, so while it looks fine, would defer to @jeremyk-91 for the approval.

**Comment by jeremyk-91** [comment]
Looks fine, but would be good to add the test that @Jolyon-S suggested. It may be a bit early, but could be worth taking a stab at adding a sanity check for the getCommitTimestamp one.

**Comment by jeremyk-91** [comment]
guess this counts, so 🕐 0:05 🏁 0:39

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java (hunk 1 above)
Comment (sudiksha27) [reply timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:503]: Leadership ids should be different as each client will have its own instance of AsyncTimelockService -> AsyncLockService -> LeaderClock. I've added the check now.

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java [hunk 2]
```diff
@@ -484,6 +491,37 @@ public void stressTestForPaxosEndpoints() {
...
+    public void leaderCanProcessMultiClientGetCommitTimestampRequest() {
+        MultiClientConjureTimelockService service =
+                cluster.currentLeaderFor(client.namespace()).multiClientService();
+        Set<String> expectedNamespaces = ImmutableSet.of("alpha", "beta");
+        List<NamespacedGetCommitTimestampsResponse> commitTimestamps =
+                service.getCommitTimestamps(AUTH_HEADER, getGetCommitTimestampsRequests(expectedNamespaces));
+        Set<String> namespaces = commitTimestamps.stream()
+                .map(NamespacedGetCommitTimestampsResponse::getNamespace)
+                .collect(Collectors.toSet());
+        assertThat(namespaces).hasSameElementsAs(expectedNamespaces);
```
Comment (Jolyon-S) [inline timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:522]: This is perhaps being a bit pedantic, but get commit timestamps have a leader ID hidden inside (it's the `logId` inside `LockWatchStateUpdate`). Perhaps worth a similar check to the one above?

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java [hunk 3]
```diff
@@ -484,6 +491,37 @@ public void stressTestForPaxosEndpoints() {
...
+                multiClientConjureTimelockService.leaderTimes(AUTH_HEADER, expectedNamespaces);
+        Set<String> namespaces =
+                leaderTimes.stream().map(NamespacedLeaderTime::getNamespace).collect(Collectors.toSet());
+        assertThat(namespaces).hasSameElementsAs(expectedNamespaces);
+
+        Set<LeadershipId> leadershipIds = leaderTimes.stream()
+                .map(NamespacedLeaderTime::getLeaderTime)
+                .map(LeaderTime::id)
+                .collect(Collectors.toSet());
+        assertThat(leadershipIds).hasSameSizeAs(expectedNamespaces);
```
Comment (jeremyk-91) [inline timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:509]: maybe a bit overkill: if we run these against the regular conjuretimelockservice.getLeaderTime we get the same answer right?

**Review by jeremyk-91 — COMMENTED** [review]
Broadly looks good. Afraid I'd like one more test, though I wouldn't mind that being in a separate PR if it's hard to write

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java (hunk 3 above)
Comment (sudiksha27) [reply timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:515]: Same leader will serve the request, but leaderTime will be different (time elapsed b/w two requests). I will add this test for commitTs request as well, where I can add stricter assertions on the response.

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java (hunk 2 above)
Comment (sudiksha27) [reply timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:528]: Done

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java (hunk 3 above)
Comment (jeremyk-91) [reply timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:513]: Yeah, makes sense. I think I intended an ID comparison, which you've added below.

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java [hunk 4]
```diff
@@ -514,12 +520,70 @@ public void leaderCanProcessMultiClientGetCommitTimestampRequest() {
...
+                    .namespacedConjureTimelockService()
+                    .getCommitTimestamps(defaultCommitTimestampRequest());
+            assertThat(namespacedGetCommitTimestampsResponse
+                            .getLockWatchUpdate()
+                            .logId())
+                    .isEqualTo(conjureTimelockServiceGetCommitTimestampResponse
+                            .getLockWatchUpdate()
+                            .logId());
+            assertThat(namespacedGetCommitTimestampsResponse.getInclusiveUpper())
+                    .isEqualTo(conjureTimelockServiceGetCommitTimestampResponse.getInclusiveLower() - 1);
```
Comment (jeremyk-91) [inline timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:583]: nit/suggestion: Usually these assertions are `assertThat(actual).isEqualTo(expected)`, so I'd probably flip these two. Also, you can use `as(String)` to say in English what the assertion is doing, which I think is useful for this later one, which verifies `timestamps are contiguously increasing per namespace if there are no elections`.

**Review by jeremyk-91 — APPROVED** [review]
Looks good - one nit left on my end.

**Pull request merged**